### DISCUSSION TITLE
Provide a implementation of `diesel::connection::Connection` for the connection type in this crate

### DISCUSSION THREAD

**Discussion by weiznich** [blackbeam/rust-mysql-simple#320] [Ideas]
As part of the [diesel](https://github.com/diesel-rs/diesel) 2.0 release we reworked large parts of diesels internals, such that it should now be possible to provide a custom `Connection` implementation for the existing diesel mysql backend outside of the main crate. Your crate would be a good candidate for this as it already provides a pure rust mysql connection implementation. I expect that there would be a certain demand in having a pure rust connection implementation supported by diesel as well.

A diesel connection implementation would enable the usage of any supported diesel feature with a connection type from your crate. This includes all of the existing query DSL, all methods to load and map results to rust struct and even the existing migration infrastructure. 

This would require implementing [`diesel::connection::Connection`](https://docs.diesel.rs/master/diesel/connection/trait.Connection.html) either directly for your connection type or indirectly via a wrapper type. Providing this implementation as part of your crate would enable a better integration with optimizations like prepared statement caching.

If you are interested in such feature I'm happy to provide guidance on writing the corresponding code. I might even be able to help writing parts of the implementation.

**Comment by blackbeam** [comment]
Hi. This sounds great.
I'll work on this as soon as my vacation is over (since May).

**Reply by weiznich** [reply]
Is there anything I could provide to make implementing this easier for you?

**Reply by blackbeam** [reply]
Couldn't say cuz did not tried yet. Too busy at the moment 🤷‍♂️

**Reply by somehowchris** [reply]
Was/Is there some progress? Would love to see this fly. Could invest some time during semester holidays in late august

**Reply by weiznich** [reply]
At least I did not have any time to work on this yet. I would be happy to see some progress here. If someone is interested in working on this I'm happy to provide some help on how to implement things so that they work well with diesel. In this case please reach out to me.

**Reply by somehowchris** [reply]
Is everything related to what’s need to be done inside [`diesel/mysql/connection`](https://github.com/diesel-rs/diesel/tree/master/diesel/src/mysql/connection)? This create already provides url parsing, statement caching, so for implementing this, the parsing of rows and returning or errors is the main thing? Just trying to get a grip for how much there is to be done for this implementation

**Reply by weiznich** [reply]
[diesel/mysql/connection](https://github.com/diesel-rs/diesel/tree/master/diesel/src/mysql/connection) provides an connection implementation based on libmysqlclient, which is the official client side c implementation of mysql's wire protocol. This crate (rust-mysql-simple) provides a pure rust implementation of the same protocol. The idea presented in the first post is to provide a `diesel::Connection` implementation based on this pure rust implementation. 
So yes this on the one hand this would implement similar code than already present in `diesel::mysql::connection`, but on the other hand it would likely be much simpler and easier to implement (as it don't need to interact with an low level C-API).

Such an implementation could be provided as a third party crate, as part of `rust-mysql-simple` or as part of `diesel`. In my role as diesel maintainer I would prefer one of the first two solutions for now.

To implement this you would need to implement `diesel::Connection` for [`mysql::Conn`](https://docs.rs/mysql/22.2.0/mysql/struct.Conn.html) (or a new type wrapper around this type) + all dependent traits (`diesel::row::Row`, …) for the corresponding types.
 
> This create already provides url parsing

I don't think this would need to handle url parsing in a special way, as this is already provided by the mysql crate (or better by [`mysql::OptsBuilder`](https://docs.rs/mysql/22.2.0/mysql/struct.OptsBuilder.html). 

> statement caching

For caching prepared statements, diesel already provides an generic implementation [here](https://docs.diesel.rs/master/diesel/connection/statement_cache/index.html). 

> the parsing of rows

I think this can be handled by "just" implementing [`diesel::row::Row`](https://docs.diesel.rs/master/diesel/row/trait.Row.html) for [`mysql::Row`](https://docs.rs/mysql/22.2.0/mysql/struct.Row.html) (or a new type wrapper). This likely requires some conversing from the value type provided by the mysql crate to the value type provided by diesel.

> returning or errors

This is currently completely missing. Likely it would require translating the error type of the mysql crate to diesel errors.

**Reply by somehowchris** [reply]
Just started out, @weiznich is there a way to split the mysql backend and mysql connection parts of diesel via features? Currently the `mysql` feature automatically enables the `mysqlclient-sys` feature

**Reply by somehowchris** [reply]
nvm, sry saw that upstream there is a `mysql_backend` feature, but as 2.0 isn't yet released, cargo add didn't show me these features

**Reply by weiznich** [reply]
Yes such an implementation should depend on diesel 2.0 and use the `mysql_backend` feature. You can use the [2.0.0-rc.1](https://crates.io/crates/diesel/2.0.0-rc.1) for the initial development. Hopefully we are able to release a final 2.0 version this month. There are no large breaking changes planed that could affect a third party connection implementation.

**Reply by somehowchris** [reply]
Is there anything against deriving `Debug` for `diesel::connection::AnsiTransactionManager`?

**Reply by weiznich** [reply]
Not really. Adding such a derive should be fine. It's likly just an oversight that it is not already there. I will open a PR for this later today.
For now you can use `TransactionManager::transaction_manager_status_mut` to get access to the only inner field. That type returned there already implements `Debug`.
